crates/re_space_view_spatial/src/space_view_2d.rs

```diff
@@ -171,91 +169,58 @@ impl SpaceViewClass for SpatialSpaceView2D {
-                        // root of the subspace, below. If there were multiple subspaces, keep them, even if
-                        // they may be redundant with the root space.
-                        if recommended_space_views.len() == 1 {
-                            recommended_space_views.clear();
-                        }
-
-                        // If this is explicitly a 2D subspace (such as from a pinhole), or there were no
-                        // other image-bucketed recommendations, create a space at the root of the subspace.
-                        if subspace.dimensionality == SubSpaceDimensionality::TwoD
-                            || recommended_space_views.is_empty()
-                        {
-                            recommended_space_views.push(RecommendedSpaceView {
-                                root: subspace.origin.clone(),
-                                query_filter: EntityPathFilter::subtree_entity_filter(
-                                    &subspace.origin,
-                                ),
-                            });
-                        }
-
-                        recommended_space_views
-                    })
-                    .collect(),
-            })
-            .unwrap_or_default();
+        SpatialTopology::access(ctx.entity_db.store_id(), |topo| SpaceViewSpawnHeuristics {
+            recommended_space_views: topo
+                .iter_subspaces()
+                .flat_map(|subspace| {
+                    if subspace.dimensionality == SubSpaceDimensionality::ThreeD
+                        || subspace.entities.is_empty()
+                        || indicated_entities.is_disjoint(&subspace.entities)
+                    {
+                        return Vec::new();
+                    }
 
-        // Find all entities that are not yet covered by the recommended space views and create a recommended
-        // space-view for each one at that specific entity path.
-        // TODO(jleibs): This is expensive. Would be great to track this as we build up the covering instead.
-        {
-            profile_scope!("space_view_2d: find uncovered entities");
-            let remaining_entities = indicated_entities
-                .iter()
-                .filter(|entity| {
-                    heuristics
-                        .recommended_space_views
+                    // Collect just the 2d-relevant entities in this subspace
+                    let relevant_entities: IntSet<EntityPath> = subspace
+                        .entities
                         .iter()
-                        .all(|r| !r.query_filter.is_included(entity))
+                        .filter(|e| indicated_entities.contains(e))
+                        .cloned()
+                        .collect();
+
+                    // For explicit 2D pinholes start at the origin, otherwise start at the common ancestor.
```

Comment:
nit: in the near future other things may make the topological space a pure 2D space - took me a bit how you infer pinholes here, but that's just the only way right now how to end up here :)